src/transaction_builder.cpp

```diff
@@ -311,3 +424,285 @@ TransactionBuilderResult TransactionBuilder::Build()
+            vpubOldProcessed = true;
+        }
+
+        CAmount jsInputValue = 0;
+        uint256 jsAnchor;
+
+        JSDescription prevJoinSplit;
+
+        // Keep track of previous JoinSplit and its commitments
+        if (mtx.vjoinsplit.size() > 0) {
+            prevJoinSplit = mtx.vjoinsplit.back();
+        }
+
+        // If there is no change, the chain has terminated so we can reset the tracked treestate.
+        if (jsChange == 0 && mtx.vjoinsplit.size() > 0) {
+            intermediates.clear();
+            previousCommitments.clear();
+        }
+
+        //
+        // Consume change as the first input of the JoinSplit.
+        //
+        if (jsChange > 0) {
+            // Update tree state with previous joinsplit
+            SproutMerkleTree tree;
+            {
+                LOCK(cs_coinsView);
+                auto it = intermediates.find(prevJoinSplit.anchor);
+                if (it != intermediates.end()) {
+                    tree = it->second;
+                } else if (!coinsView->GetSproutAnchorAt(prevJoinSplit.anchor, tree)) {
+                    throw JSDException("Could not find previous JoinSplit anchor");
+                }
+            }
+
+            assert(changeOutputIndex != -1);
+            assert(changeOutputIndex < prevJoinSplit.commitments.size());
+            boost::optional<SproutWitness> changeWitness;
+            int n = 0;
+            for (const uint256& commitment : prevJoinSplit.commitments) {
+                tree.append(commitment);
+                previousCommitments.push_back(commitment);
+                if (!changeWitness && changeOutputIndex == n++) {
+                    changeWitness = tree.witness();
+                } else if (changeWitness) {
+                    changeWitness.get().append(commitment);
+                }
+            }
+            assert(changeWitness.has_value());
+            jsAnchor = tree.root();
+            intermediates.insert(std::make_pair(tree.root(), tree)); // chained js are interstitial (found in between block boundaries)
+
+            // Decrypt the change note's ciphertext to retrieve some data we need
+            ZCNoteDecryption decryptor(changeKey.receiving_key());
+            auto hSig = prevJoinSplit.h_sig(*sproutParams, mtx.joinSplitPubKey);
+            try {
+                auto plaintext = libzcash::SproutNotePlaintext::decrypt(
+                    decryptor,
+                    prevJoinSplit.ciphertexts[changeOutputIndex],
+                    prevJoinSplit.ephemeralKey,
+                    hSig,
+                    (unsigned char)changeOutputIndex);
+
+                auto note = plaintext.note(changeAddress);
+                vjsin[0] = libzcash::JSInput(changeWitness.get(), note, changeKey);
+
+                jsInputValue += plaintext.value();
+
+                LogPrint("zrpcunsafe", "spending change (amount=%s)\n", FormatMoney(plaintext.value()));
+
+            } catch (const std::exception& e) {
```

Comment:
Narrow this try-catch block if possible, since the error message assumes that the exception was thrown from `libzcash::SproutNotePlaintext::decrypt`. Non-blocking.

Reply:
We are doing essentially the same thing in `z_mergetoaddress` and `z_sendmany`, so I am happy leaving it as it for the sake of consistency. That being said, I would be in favor of moving the catch block to just after we call decrypt, and then expanding the try/catch where we call `CreateJSDescriptions` to catch and rethrow std::exception in addition to `JSDException` and `std::runtime_error`. What do you think of that?

Reply:
Please file a ticket to do this, and also to update `z_mergetoaddress` and `z_sendmany` in the same way.

Reply:
https://github.com/zcash/zcash/issues/3904